DIPs/DIP1NNN-DK.md

```diff
@@ -0,0 +1,696 @@
+
+### Proposed changes
+
+**(0) _Writing_ to variables or fields marked `@system` is not allowed in `@safe` code**
+
+This does not apply to local variables however, where `@system` remains ignored.
+A function already has full control over the scope of its local variables, unless a `@safe` way of inspecting closures is introduced.
+
+```D
+@system int x;
+
+struct S {
+    @system int y;
+}
+
+S s;
+
+void main() @safe {
+    x += 10; // error: can't modify @system variable 'x'
+    s.y += 10; // error: can't modify @system field 'y'
+
+    @system int z;
+    z += 1; // still allowed
+}
+
+// inferred as a @system function
+auto foo() {
+    x = 0;
+}
+```
+
+Further operations disallowed in `@safe` code on `@system` variables or fields are:
+- taking its address using `&`
+- passing it as an argument to a function parameter marked `ref`
+- returning it by `ref`
+
+A struct or union with `@system` fields can not be constructed using the default constructor in `@safe` code.
+A `@trusted` constructor can be defined to allow construction in `@safe` code.
+Note that the `.init` value of a type is always usable in `@safe` code.
+
+When using an `alias` to a `@system` field, that alias has the same restrictions regarding `@system`.
+
+**(1) An aggregate with at least one `@system` field is an unsafe type**
+
+It gets the same restrictions as existing unsafe types:
+
+```D
+struct Handle {
+    @system int handle;
+}
+
+void main() @safe {
+    Handle h = void; // error
+    union U {
+        Handle h;
+        int i;
+    }
+    U u;
+    u.i = 3; // error
+
+    ubyte[Handle.sizeof] storage;
+    auto array = cast(Handle[]) storage[]; // error
+}
+```
+
+Without this, implicit writes to `@system` variables are still possible.
+
+**(2) _Reading_ from variables or fields marked `@system` is not allowed in `@safe` code if their type is unsafe**
+
+While writing to a `@system` variable is always unsafe, reading from one is only dangerous when it could yield an unsafe value.
+
+```D
+struct Handle {
+    @system int handle;
+}
+
+// struct with @system field is an unsafe type
+@safe   Handle safeHandle = Handle(1);
+@system Handle systemHandle = Handle(-1);
+
+// pointers are an unsafe type
+@safe   immutable int* safePtr   = null;
+@system immutable int* systemPtr = cast(int*) 0x8035FDF0;
+
+// integers are a safe type
+@safe   int safeInt   = 20;
+@system int systemInt = 20;
+
+void main() @safe {
+    Handle h0 = safeHandle;        // allowed, @safe variable
+    Handle h1 = systemHandle;      // error, reading @system var of unsafe type
+    immutable int* p0 = safePtr;   // allowed, @safe variable
+    immutable int* p1 = systemPtr; // error, reading @system var of unsafe type
+    int i0 = safeInt;              // allowed
+    int i1 = systemInt;            // allowed, not an unsafe type
+}
+```
+
+**(3) Variables and fields without annotation are `@safe` unless their initial value is not `@safe`**
```

Comment:
What's the rationale for prohibiting reading such values?

Reply:
To prevent memory corruption. If you could read `systemPtr`, you can dereference it an access an arbitrary memory address.

Reply:
Reading the pointer is not the same thing as dereferencing it.

Reply:
The thing is that `@system` is not part of the type like `const` or `shared`, so you can easily strip off `@system` by simply reading the value, after which you can dereference the garbage pointer:
```D
@system int* garbage = cast(int*) 0xDEADBEEF;

void main() @safe {
    *garbage = 0; // error! this will be caught
    int* pointer = garbage; // this is what rule (3) disallows
    *pointer = 0; // this is allowed, so the above line should be prevented
```